src/hooks/progress/use-update-tutorial-progress.ts

```diff
@@ -0,0 +1,186 @@
+ *
+ * When the endRef element moves into the viewport,
+ * the tutorial will be marked as `complete` (100 percent).
+ */
+export function useUpdateTutorialProgress({
+	tutorialId,
+	collectionId,
+}: TutorialIdCollectionId): UpdateProgressReturnInterface {
+	// We shouldn't try to update progress unless we're authenticated
+	const { isAuthenticated } = useAuthentication()
+	// We need to know if progress exists, to know whether to "create" or "update"
+	const { tutorialProgressStatus } = useTutorialProgress({
+		tutorialId,
+		collectionId,
+	})
+	// We'll use "create" or "update" progress using these `react-query` mutations
+	const tutorialProgressMutations = useTutorialProgressMutations()
+	// We use refs from `react-intersection-observer` to track progress.
+	const [startRef, startInView, startEntry] = useInView({
+		initialInView: true,
+	})
+	const [endRef, endInView, endEntry] = useInView()
+
+	/**
+	 * This effect runs often, but it seems necessary to ensure progress
+	 * gets tracked as things like the authenticated state changes.
+	 *
+	 * We bail early in many cases, and we avoid downgrading progress,
+	 * so this effect only results in meaningful mutations a small percent
+	 * of the time.
+	 */
+	useEffect(() => {
+		/**
+		 * On initial client-side navigation, this effect may be called
+		 * before startRefs and endRefs are updated. We can detect this
+		 * by comparing data-ref-id attributes we've added to the
+		 * elements startRef and endRef are attached to.
+		 */
+		const startRefId = startEntry?.target.getAttribute('data-ref-id')
+		const endRefId = endEntry?.target.getAttribute('data-ref-id')
+		const compareId = `${tutorialId}_${collectionId}`
+		const isStaticIdsMismatch =
+			startRefId !== compareId || endRefId !== compareId
+
+		console.log({
+			isStaticIdsMismatch,
+			startInView,
+			endInView,
+			startRefId,
+			endRefId,
+			compareId,
+		})
+		if (isStaticIdsMismatch) {
+			return
+		}
+
+		/**
+		 * If we're not authenticated, we bail early, there's no
+		 * point in trying to make a mutation.
+		 */
+		if (!isAuthenticated) {
+			return
+		}
+
+		/**
+		 * If we are authenticated but we don't have a result
+		 * yet from querying this tutorial's status, we also bail early.
+		 * We need that result to know whether to "create" or "update".
+		 */
+		if (typeof tutorialProgressStatus == 'undefined') {
+			return
+		}
+
+		/**
+		 * We don't want to run multiple mutations at once.
+		 * To avoid this, tutorialProgressMutations provided `isReady`,
+		 * which we respect here.
+		 */
+		const isReadyToRunMutation = tutorialProgressMutations.isReady
+		if (!isReadyToRunMutation) {
+			return
+		}
+
+		/**
+		 * Determine the new progress state, which we'll then update
+		 * - 'complete' (100%) if the end ref is visible
+		 * - 'in_progress' (50%) if the start ref is no longer visible
+		 * - 'visited' (0%) in all other cases
+		 */
+		let newProgressStatus: TutorialProgressStatus
```

Comment:
💭 It might help with readability to encapsulate this into a function. So one can easily scan all the early returns and then read name like `handleUpdateProgressStatus` to make this useEffect easy to digest. 

Just a thought. It's up to you.

Reply:
I'm in agreement here; this `useEffect` is quite long at over 100 lines. I think we could  quickly break it into a couple of helpers. 😊

- `canUpdateTutorialProgress`: this would contain all of the early bail `if` conditions that just do `return` when the conditions are true.
- `updateTutorialProgress`: this would run the code within the `if (needsInitialProgress || hasPositiveProgress)` block.

Feel free to use any other function names! These are just my quick ideas for naming.

Reply:
✅ updated to split out `canUpdateTutorialProgress` & `updateTutorialProgress` (names are great, hardest problem solved haha 🙌 )